src/Axe/Controllers/CompilerController.cs

```diff
@@ -19,13 +19,32 @@
 using Microsoft.AspNetCore.NodeServices;
 using Microsoft.AspNetCore.Mvc.Rendering;
 using Microsoft.AspNetCore.Authorization;
+using System.Net.WebSockets;
+using Microsoft.AspNetCore.Http;
+using System.Threading;
+using Newtonsoft.Json;
+using System.Text;
+using Axe.Dto.Compiler;
+using Newtonsoft.Json.Serialization;
 
 namespace Axe.Controllers
 {
     public class CompilerController : ControllerExt
     {
-        private ICompilerManager compileManager;
-        private ITechnologyManager technologyManager;
+        #region FIELDS
+        private readonly ICompilerManager compileManager;
+        private readonly ITechnologyManager technologyManager;
+
+        /// <summary>
+        /// Dictionary of observers
+        /// </summary>
+        private static Dictionary<string, WebSocketCompilerWrapper> observersWebSockets = new Dictionary<string, WebSocketCompilerWrapper>();
+        
+        /// <summary>
+        /// Dictionary of users which try to solve tasks
+        /// </summary>
+        private static Dictionary<string, WebSocketCompilerWrapper> usersWebSockets = new Dictionary<string, WebSocketCompilerWrapper>();
```

Comment:
CompilerController is too heavy. It contains a lot of business logic. Business logic should be extracted in a separate class (which can be properly tested). Controller will handle rounting and pass request->manager/response->client.